File: misc/load_db.py
#!/usr/bin/env python3
"""Load enriched student data from newly_scraped table into person table."""

import os
import re
import html as html_lib
import sqlalchemy
from sqlalchemy import text
from dotenv import load_dotenv

load_dotenv()

DATABASE_URL = os.environ["DATABASE_URL"]

MONTH_MAP = {
    "Jan": 1, "Feb": 2, "Mar": 3, "Apr": 4, "May": 5, "Jun": 6,
    "Jul": 7, "Aug": 8, "Sep": 9, "Oct": 10, "Nov": 11, "Dec": 12,
}


def parse_birthday(bday_str):
    """Parse 'Dec 29' into (month_int, day_int)."""
    if not bday_str:
        return None, None
    match = re.match(r"(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s+(\d{1,2})", bday_str)
    if match:
        return MONTH_MAP[match.group(1)], int(match.group(2))
    return None, None


def parse_year(year_str, year_dir):
    """Convert year to int. Prefer directory year, fallback to facebook year."""
    if year_dir:
        return int(year_dir)
    if year_str:
        match = re.match(r"'(\d{2})", year_str)
        if match:
            return 2000 + int(match.group(1))
    return None


def clean_str(value):
    """Clean a string value: decode HTML entities, strip whitespace."""
    if not value or not str(value).strip():
        return None
    return html_lib.unescape(str(value).strip())


def to_db_row(student):
    """Convert enriched student dict to a database row dict matching person table."""
    birth_month, birth_day = parse_birthday(student.get("birthday"))
    year = parse_year(student.get("year"), student.get("year_directory"))

    upi = student.get("upi")
    photo_id = student.get("photo_id")

    # Build image URL from photo_id (served from GCS bucket)
    image = None
    if photo_id and photo_id != "0":
        image = f"https://storage.googleapis.com/yalies-photos/{photo_id}.jpg"

    # Build address: prefer directory student_address, fallback to scraped
    address = student.get("student_address") or student.get("address", "")
    if address:
        address = address.replace(" | ", "\n")

    return {
        "id": upi if upi else None,
        "netid": clean_str(student.get("netid")),
        "upi": upi,
        "email": clean_str(student.get("email")),
        "mailbox": clean_str(student.get("mailbox")),
        "phone": clean_str(student.get("phone_directory") or student.get("phone")),
        "first_name": clean_str(student.get("first_name_directory") or student.get("first_name")) or "",
        "preferred_name": clean_str(student.get("preferred_name")),
        "middle_name": clean_str(student.get("middle_name")),
        "last_name": clean_str(student.get("last_name")) or "",
        "suffix": clean_str(student.get("suffix")),
        "pronouns": clean_str(student.get("pronouns")),
        "school": clean_str(student.get("school")) or "Yale College",
        "school_code": clean_str(student.get("school_code")) or "YC",
        "year": year,
        "curriculum": clean_str(student.get("curriculum")),
        "college": clean_str(student.get("college_directory") or student.get("college")),
        "college_code": clean_str(student.get("college_code")),
        "image": image,
        "birth_month": birth_month,
        "birth_day": birth_day,
        "major": clean_str(student.get("major")),
        "address": clean_str(address),
        "organization": clean_str(student.get("organization")),
        "organization_code": clean_str(student.get("organization_code")),
        "unit": clean_str(student.get("unit")),
        "postal_address": clean_str(student.get("postal_address")),
    }


def load_from_newly_scraped(engine):
    """Load all rows from newly_scraped table as dicts."""
    with engine.connect() as conn:
        rows = conn.execute(text("SELECT * FROM newly_scraped ORDER BY id")).fetchall()
    return [dict(row._mapping) for row in rows]


def main():
    import argparse
    parser = argparse.ArgumentParser(description="Load students into database")
    parser.add_argument("--database-url", type=str, default=DATABASE_URL,
                        help="PostgreSQL connection URL")
    parser.add_argument("--dry-run", action="store_true",
                        help="Show what would happen without modifying the database")
    args = parser.parse_args()

    engine = sqlalchemy.create_engine(args.database_url)

    students = load_from_newly_scraped(engine)
    print(f"Loaded {len(students)} students from newly_scraped table")

    with engine.begin() as conn:
        # Pre-check: count existing Yale College rows
        result = conn.execute(text(
            "SELECT COUNT(*) FROM person WHERE school = 'Yale College' OR school_code = 'YC'"
        ))
        existing_count = result.scalar()
        print(f"Existing Yale College rows: {existing_count}")

        if args.dry_run:
            print(f"DRY RUN: Would delete {existing_count} rows and insert ~{len(students)}")
            return

        # Step 1: Delete all existing Yale College rows
        result = conn.execute(text(
            "DELETE FROM person WHERE school = 'Yale College' OR school_code = 'YC'"
        ))
        print(f"Deleted {result.rowcount} existing Yale College rows")

        # Step 2: Insert all new students
        inserted = 0
        skipped = 0
        seen_ids = set()
        id_conflicts = []

        for student in students:
            row = to_db_row(student)

            # Generate a stable ID for students without UPI
            if not row["id"]:
                # Use a deterministic hash based on name + year to avoid collisions
                name_key = f"YC_{row['first_name']}_{row['last_name']}_{row.get('year', '')}_{row.get('college', '')}"
                row["id"] = abs(hash(name_key)) % (2**31 - 1)

            # Handle duplicate IDs
            if row["id"] in seen_ids:
                # Append a counter to make unique
                base_id = row["id"]
                counter = 1
                while row["id"] in seen_ids:
                    row["id"] = (base_id + counter) % (2**31 - 1)
                    counter += 1
                id_conflicts.append(f"{row['first_name']} {row['last_name']}")

            seen_ids.add(row["id"])

            # Filter out None values for the insert
            columns = [k for k, v in row.items() if v is not None]
            values = {k: v for k, v in row.items() if v is not None}
            placeholders = ", ".join(f":{k}" for k in columns)
            col_names = ", ".join(columns)

            conn.execute(
                text(f"INSERT INTO person ({col_names}) VALUES ({placeholders})"),
                values
            )
            inserted += 1

        print(f"Inserted {inserted} students")
        if id_conflicts:
            print(f"Resolved {len(id_conflicts)} ID conflicts: {id_conflicts[:5]}...")

        # Step 3: Verify
        result = conn.execute(text("SELECT COUNT(*) FROM person"))
        total = result.scalar()
        result = conn.execute(text("SELECT COUNT(*) FROM person WHERE school = 'Yale College'"))
        yc = result.scalar()
        result = conn.execute(text("SELECT COUNT(*) FROM person WHERE school = 'Yale College' AND netid IS NOT NULL"))
        with_netid = result.scalar()

        print(f"\nDatabase summary:")
        print(f"  Total rows: {total}")
        print(f"  Yale College rows: {yc}")
        print(f"  Yale College with netid: {with_netid}")
        print(f"  Non-Yale College rows: {total - yc}")


if __name__ == "__main__":
    main()

File: Yalies/yalies-internal/yalies-data-pipeline/src/loadDb.ts
import { Sequelize, QueryTypes } from "sequelize";
import { decode } from "html-entities";
import { EnrichedStudent, DbRow } from "./types.js";
import { parseLocation } from "./parseLocation.js";
import { YALE_COLLEGE, YALE_COLLEGE_CODE } from "yalies-shared";

const MONTH_MAP: Record<string, number> = {
	"Jan": 1, "Feb": 2, "Mar": 3, "Apr": 4, "May": 5, "Jun": 6,
	"Jul": 7, "Aug": 8, "Sep": 9, "Oct": 10, "Nov": 11, "Dec": 12,
};

function parseBirthday(bday?: string): { month?: number; day?: number } {
	if (!bday) return {};
	const match = bday.match(/^(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\s+(\d{1,2})$/);
	if (match) {
		return { month: MONTH_MAP[match[1]], day: parseInt(match[2]) };
	}
	return {};
}

function parseYear(yearStr?: string, yearDir?: number): number | undefined {
	if (yearDir) return yearDir;
	if (yearStr) {
		const match = yearStr.match(/^'(\d{2})$/);
		if (match) return 2000 + parseInt(match[1]);
	}
	return undefined;
}

function cleanStr(value?: string | number | null): string | undefined {
	if (value === null || value === undefined) return undefined;
	const str = decode(String(value).trim());
	return str.length > 0 ? str : undefined;
}

export function toDbRow(student: EnrichedStudent): DbRow {
	const { month: birthMonth, day: birthDay } = parseBirthday(student.birthday);
	const year = parseYear(student.year, student.year_directory);
	const upi = student.upi;
	const photoId = student.photo_id;

	let image: string | undefined;
	if (photoId && photoId !== "0") {
		image = `https://storage.googleapis.com/yalies-photos/${photoId}.jpg`;
	}

	let address = student.student_address || student.address || "";
	if (address) {
		address = address.replace(/ \| /g, "\n");
	}

	const cleanedAddress = cleanStr(address);
	const { address_state, address_country } = parseLocation(cleanedAddress);

	return {
		id: upi || 0,
		netid: cleanStr(student.netid)?.toLowerCase(),
		upi,
		email: cleanStr(student.email),
		mailbox: cleanStr(student.mailbox),
		phone: cleanStr(student.phone_directory || student.phone),
		first_name: cleanStr(student.first_name_directory || student.first_name) || "",
		preferred_name: cleanStr(student.preferred_name),
		middle_name: cleanStr(student.middle_name),
		last_name: cleanStr(student.last_name) || "",
		suffix: cleanStr(student.suffix),
		pronouns: cleanStr(student.pronouns),
		school: cleanStr(student.school) || YALE_COLLEGE,
		school_code: cleanStr(student.school_code) || YALE_COLLEGE_CODE,
		year,
		curriculum: cleanStr(student.curriculum),
		college: cleanStr(student.college_directory || student.college),
		college_code: cleanStr(student.college_code),
		image,
		birth_month: birthMonth,
		birth_day: birthDay,
		major: cleanStr(student.major),
		address: cleanedAddress,
		address_state: address_state ?? undefined,
		address_country: address_country ?? undefined,
		organization: cleanStr(student.organization),
		organization_code: cleanStr(student.organization_code),
		unit: cleanStr(student.unit),
		postal_address: cleanStr(student.postal_address),
	};
}

export async function loadToDatabase(
	students: EnrichedStudent[],
	databaseUrl: string,
	dryRun = false,
	force = false,
): Promise<void> {
	console.log(`Loading ${students.length} students into database...`);

	const sequelize = new Sequelize(databaseUrl, { logging: false });
	const ycReplacements = { yc: YALE_COLLEGE, ycCode: YALE_COLLEGE_CODE };

	try {
		await sequelize.authenticate();
		console.log("Connected to database");

		// Pre-check
		const [existingResult] = await sequelize.query<{ count: string }>(
			"SELECT COUNT(*) as count FROM person WHERE school = :yc OR school_code = :ycCode",
			{ type: QueryTypes.SELECT, replacements: ycReplacements },
		);
		const existingCount = parseInt(existingResult.count);
		console.log(`Existing ${YALE_COLLEGE} rows: ${existingCount}`);

		if (dryRun) {
			console.log(`DRY RUN: Would delete ${existingCount} rows and insert ~${students.length}`);
			return;
		}

		// Safety guard: refuse to replace a healthy roster with a much smaller
		// one (e.g. a partial scrape from an expired cookie) unless forced.
		// Runs after the dry-run return so a preview is never blocked.
		if (!force && existingCount > 0 && students.length < existingCount * 0.8) {
			throw new Error(
				`Refusing to sync: new set (${students.length}) is under 80% of existing ` +
				`${YALE_COLLEGE} rows (${existingCount}). Re-run with --force to override.`,
			);
		}

		const transaction = await sequelize.transaction();

		try {
			// Step 1: Delete existing rows
			await sequelize.query(
				"DELETE FROM person WHERE school = :yc OR school_code = :ycCode",
				{ transaction, replacements: ycReplacements },
			);
			console.log(`Deleted existing ${YALE_COLLEGE} rows`);

			// Step 2: Insert new students
			let inserted = 0;
			const seenIds = new Set<number>();
			const idConflicts: string[] = [];

			for (const student of students) {
				const row = toDbRow(student);

				// Generate stable ID for students without UPI
				if (!row.id) {
					const nameKey = `${YALE_COLLEGE_CODE}_${row.first_name}_${row.last_name}_${row.year || ""}_${row.college || ""}`;
					row.id = Math.abs(hashCode(nameKey)) % (2 ** 31 - 1);
				}

				// Handle duplicate IDs
				if (seenIds.has(row.id)) {
					const baseId = row.id;
					let counter = 1;
					while (seenIds.has(row.id)) {
						row.id = (baseId + counter) % (2 ** 31 - 1);
						counter++;
					}
					idConflicts.push(`${row.first_name} ${row.last_name}`);
				}
				seenIds.add(row.id);

				// Build insert query with only non-null values
				const entries = Object.entries(row).filter(([, v]) => v !== undefined && v !== null);
				const columns = entries.map(([k]) => k).join(", ");
				const placeholders = entries.map(([k]) => `:${k}`).join(", ");
				const replacements = Object.fromEntries(entries);

				await sequelize.query(
					`INSERT INTO person (${columns}) VALUES (${placeholders})`,
					{ replacements, transaction },
				);
				inserted++;
			}

			await transaction.commit();

			console.log(`Inserted ${inserted} students`);
			if (idConflicts.length > 0) {
				console.log(`Resolved ${idConflicts.length} ID conflicts: ${idConflicts.slice(0, 5).join(", ")}...`);
			}

			// Step 3: Verify
			const [totalResult] = await sequelize.query<{ count: string }>(
				"SELECT COUNT(*) as count FROM person",
				{ type: QueryTypes.SELECT },
			);
			const [ycResult] = await sequelize.query<{ count: string }>(
				"SELECT COUNT(*) as count FROM person WHERE school = :yc",
				{ type: QueryTypes.SELECT, replacements: ycReplacements },
			);
			const [netidResult] = await sequelize.query<{ count: string }>(
				"SELECT COUNT(*) as count FROM person WHERE school = :yc AND netid IS NOT NULL",
				{ type: QueryTypes.SELECT, replacements: ycReplacements },
			);

			console.log("\nDatabase summary:");
			console.log(`  Total rows: ${totalResult.count}`);
			console.log(`  ${YALE_COLLEGE} rows: ${ycResult.count}`);
			console.log(`  ${YALE_COLLEGE} with netid: ${netidResult.count}`);
			console.log(`  Non-${YALE_COLLEGE} rows: ${parseInt(totalResult.count) - parseInt(ycResult.count)}`);
		} catch (e) {
			await transaction.rollback();
			throw e;
		}
	} finally {
		await sequelize.close();
	}
}

function hashCode(str: string): number {
	let hash = 0;
	for (let i = 0; i < str.length; i++) {
		const char = str.charCodeAt(i);
		hash = ((hash << 5) - hash) + char;
		hash |= 0; // Convert to 32-bit integer
	}
	return hash;
}
